coopr-server/src/main/java/co/cask/coopr/upgrade/UpgradeTo0_9_9.java

```diff
@@ -76,85 +77,124 @@ public void run(long batchSize) throws IOException, SQLException {
     denormalizeTasksTable();
   }
 
+  private int getTableRowCount(Connection conn, String table) throws SQLException {
+    PreparedStatement statement = conn.prepareStatement("SELECT COUNT(*) FROM " + table);
+    try {
+      ResultSet rs = statement.executeQuery();
+      try {
+        if (rs.next()) {
+          return rs.getInt(1);
+        }
+      } finally {
+        rs.close();
+      }
+    } finally {
+      statement.close();
+    }
+
+    throw new IllegalStateException();
+  }
+
   private void denormalizeTasksTable() throws IOException, SQLException {
     LOG.info("Denormalizing tasks table: type, cluster_template_name, user_id, and tenant_id columns");
+
+    LOG.info("Fetching all clusters");
     List<Cluster> clusters = clusterStore.getAllClusters();
     Map<Long, Cluster> clustersById = Maps.newHashMap();
     for (Cluster cluster : clusters) {
       clustersById.put(Long.parseLong(cluster.getId()), cluster);
     }
 
+    ArrayList<Task> tasks = null;
     Connection conn = dbConnectionPool.getConnection();
     try {
-      PreparedStatement statement = conn.prepareStatement("SELECT * FROM tasks");
+      int approxNumTasks = getTableRowCount(conn, "tasks");
+      tasks = new ArrayList<Task>(approxNumTasks);
+
+      // populate tasks
+      LOG.info("Fetching all tasks");
+      PreparedStatement selectAllTasks = conn.prepareStatement("SELECT * FROM tasks");
+      try {
+        ResultSet rs = selectAllTasks.executeQuery();
```

Comment:
we shouldn't load all tasks in memory, there can be quite a lot.  What you could do is get the max task id, then do tasks 0 - 1000, then 1001 - 2000, etc.